### ISSUE TITLE
Failed merging two backups :-(

### ISSUE THREAD

**Issue by fattybenji** [bepaald/signalbackup-tools#173]
Writing this as feedback since you asked :-)

I installed the tool today, and I used a pretty old backup as the first one and a pretty recent as the second one. Some numbers because we like numbers:

* Database versions were 199 and 211 (probably the issue here)
* I used sqlite 3.43.2
* I'm on macOS Sonoma 14.2.1

The error I got was this:

```
[Error]: During sqlite3_prepare_v2(): no such column: uuid
         -> Query: "SELECT IFNULL(uuid, '') AS uuid, IFNULL(phone, '') AS phone, IFNULL(group_id, '') AS group_id FROM recipient WHERE _id IS (SELECT recipient_id FROM thread WHERE _id = ?)"
Failed to get uuid/phone/group_id from source database
[Error]: A fatal error occurred while trying to import thread 1. Aborting
```

It would be nice to have a "migration tool" so we can get around this (but if you can migrate you can do this right away I guess)

**Comment by bepaald** [comment]
Hi! Thanks for reporting! The database versions here are indeed the issue. Is it not possible for you to import the 199 backup into Signal and export it again? That would turn it into a 211 version.

The versions you are using are not too far apart, just a bit unlucky around the renaming of `uuid` (in 200). If you have no other options, I _might_ be able to work around this issue. But that could take a couple of days, let me know.

Thanks again!

**Comment by fattybenji** [comment]
Thanks for the answer ! Unfortunately, the old backup is too big for Signal to handle (2+GB) so I was not able to import it. It crashed my older phone, and on a newer one Signal just stops unexpectedly. I could maybe slice it up but it's quite an annoying process 😅 

Too bad for me, I'll just not have those messages 🤷

**Comment by bepaald** [comment]
I made an attempt to allow merging database version 199 and 211 (or higher). I am not sure if it works, but you could certainly update and try if you want to.

I must say, I of course don't know what phones you are using, but 2GB is not that big. My normal backup is about 7GB and people I've spoken to (through this issue tracker) have backup files sometimes over 30GB in size. Signal should usually be able to handle that fine. It makes me think maybe something else is wrong with the old backup?

Anyway, let me know if you want to try the merge again, I'd be curious to hear the results. Otherwise, this issue can be closed.

Thanks!

**Comment by fattybenji** [comment]
I updated and tested the merge, it went further than before but broke with a foreign key error:

```
ERROR Foreign key constraint violated. This will not end well, aborting.
      Please report this error to the program author.
-----------------------------
| table    | parent  | fkid |
-----------------------------
| reaction | message | 1    |
-----------------------------
[Error]: A fatal error occurred while trying to import thread 16. Aborting
```

However, when looking at the full log, I see a whoooole bunch of errors (but maybe non-breaking ?). I got about 4k times this error:

```
[Error]: During sqlite3_prepare_v2(): no such column: group_type
         -> Query: "DELETE FROM recipient WHERE ((aci IS NOT NULL AND aci IS ?) OR (e164 IS NOT NULL AND e164 IS ?) OR (group_id IS NOT NULL AND group_id IS ?)) OR (aci IS NULL AND e164 IS NULL AND group_id IS NULL AND group_type IS 4  AND group_type IS ? AND storage_service_id IS ?)"
```

When trying to insert a recipient, I get this (but with several unique constraints failed):

```
[Error]: After sqlite3_step(): UNIQUE constraint failed: recipient.phone
         -> Query: "INSERT INTO recipient (_id,phone,uuid,pni,username,email,group_id,distribution_list_id,call_link_room_id,registered,unregistered_timestamp,blocked,hidden,profile_key,profile_key_credential,profile_sharing,signal_profile_name,profile_family_name,profile_joined_name,last_profile_fetch,system_given_name,system_family_name,system_display_name,system_nickname,system_photo_uri,system_phone_label,system_phone_type,system_contact_uri,system_info_pending,notification_channel,message_ringtone,message_vibrate,call_ringtone,call_vibrate,mute_until,message_expiration_time,storage_service_key,mention_setting,capabilities,last_session_reset,wallpaper,about,about_emoji,extras,groups_in_common,color,chat_colors,custom_chat_colors_id,badges,needs_pni_signature,reporting_token) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"
```

And then I get a few errors like this one:

```
[Error]: After sqlite3_step(): UNIQUE constraint failed: message.date_sent, message.from_recipient_id, message.thread_id
         -> Query: "INSERT INTO message (_id,date_sent,date_received,date_server,thread_id,from_recipient_id,from_device_id,to_recipient_id,type,body,read,ct_l,exp,m_type,m_size,st,tr_id,subscription_id,receipt_timestamp,delivery_receipt_count,read_receipt_count,viewed_receipt_count,mismatched_identities,network_failures,expires_in,expire_started,notified,quote_id,quote_author,quote_body,quote_missing,quote_mentions,quote_type,shared_contacts,unidentified,link_previews,view_once,reactions_unread,reactions_last_seen,remote_deleted,mentions_self,notified_timestamp,server_guid,message_ranges,story_type,parent_story_id,export_state,exported,scheduled_date,latest_revision_id,original_message_id,revision_number) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"
```

Since they didn't break the execution, they might be caught and just printed ?

About my phone, I tested the first time with a Pixel 3a (which was the one who crashed, 4GB of RAM) and then with the Fairphone 5 (8GB of RAM, the app crashed but not the phone). It might have crashed because the backup was too old and they weren't able to read the database without the uuids ? It seems weird to me because it looks like the restoring process manages to load a lot of messages and breaks towards the end.

Thank you very much for your (very) fast feedback, and if you think those error messages are just normal and my backup is lost, I'll just close this.

**Comment by bepaald** [comment]
Sorry, I think the fatal error is a result of the other errors, which were because I missed a spot in my last commit. Hopefully it's fixed now, so you could try again if you want. (though more errors may come up if I made another mistake, please let me know)

I'm not sure about this one: `UNIQUE constraint failed: message.date_sent, message.from_recipient_id, message.thread_id`. Maybe it is also the result of that one error and it's now fixed, but: Is there any overlap in the two backups you are merging? Signal backups do not allow doubled messages, if the two backups contain (partly) the same messages, you must first use `--croptodates` on one of them to remove the messages that exist in both backups.

>Thank you very much for your (very) fast feedback, and if you think those error messages are just normal and my backup is lost, I'll just close this.

I don't think that yet. So we can keep trying for now (if you still want to of course).

If you try to do `--exporthtml` on the older database, does it work? Does it print any errors?

**Comment by bepaald** [comment]
Hi! Have you given up on this already? No problem of course, but I'd just like to know. If this is not going to be tested, this issue might as well be closed. If you do still plan on testing, that's great of course and no hurry (but let me know).

Do note: since database version 215 things have changed even more, so merging older versions with 215 or higher is not (going to be) possible.

Thanks!

**Comment by fattybenji** [comment]
Hello, sorry I didn't answer for this long !

TLDR I tried again a few times but ran into multiple issues so I think we'll close that, sorry...

There was some overlap between the backups so I did cut some stuff.

With the new version, I still had the same fatal error with the foreign key, and when running `exporthtml` it worked but the output reads at one point:

```
Dealing with thread 16
[Error]: Failed to open backup file '' for reading attachment
```

Now since the errors appeared both time in thread 16 there must have been something corrupt there. I've tried cropping this thread (which is not very important) but then the same fatal error as we had above for other threads, which are more important. So I tried cropping as advised, and I'm still getting the same kind of fatal error:

```
Please report this error to the program author.
-----------------------------
| table    | parent  | fkid |
-----------------------------
| reaction | message | 1    |
-----------------------------
[Error]: A fatal error occurred while trying to import thread 24. Aborting
```

However it's with another thread again, so... I tried just skipping that one, even if I'd have liked to keep it. And I got another error !

```
[Error]: Failed to get uuid/phone/group_id from source database
[Error]: A fatal error occurred while trying to import thread 33. Aborting
```

So I tried looking at what thread 33 was and surprise, it doesn't seem to exist ! There might be some issues with the thread selector, probably something with the source database not having the right info or something

Anyway, I think it will be too difficult to actually do something there so thank you for your help, and your tool, but I've accepted my fate^^

**Comment by bepaald** [comment]
Thank you for your quick response. I'm sorry merging didn't work out for you.

While I agree closing this issue is the best option, as changes will not be tested anymore, I will take this latest info you provided to look for problems in the code and improve the program this weekend. Though, with the errors reported, I must also consider the possibility that something was not right with the backup to begin with (as restoring it failed on multiple phones before).

Thanks again!
